### PR TITLE
Enable SPMD based FSDP

### PR DESCRIPTION
**Pull request by baoleai**

### CODE REVIEW

File: examples/llama3_acc.sh [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #!/bin/bash
 set -ex
 
+rm -rf compiled_cache/
```
Comment (yitongh) [inline examples/llama3_acc.sh]: why remove this？

File: examples/run.sh [hunk 2]
```diff
@@ -10,6 +10,7 @@ NUM_EPOCHS=1        # number epoches
 MAX_STEPS=-1        # max steps
 GA=1                # gradients accumulation number
 LOG_INTERVAL=1      # log interval
+LOG_LOSS=0          # log loss
```
Comment (yitongh) [inline examples/run.sh]: You can directly run ./examples/run.sh --profile --log_loss --spmd_fsdp. The parsing of these parameters is unnecessary.

File: flashmodels/accelerators/acc_baichuan_accelerator.py [hunk 3]
```diff
@@ -17,7 +16,7 @@ def accelerate_internal(self, model, loader):
                 raise NotImplementedError("resume_from_checkpoint.")
 
             config = self.get_config(model)
```
Comment (yitongh) [inline flashmodels/accelerators/acc_baichuan_accelerator.py:19]: can we support spmd_fsdp for other models?

File: flashmodels/patch/patch.py [hunk 4]
```diff
@@ -41,18 +41,17 @@ def patch_llama():
         transformers.models.llama.modeling_llama.LlamaAttention.forward = flash_attn_fwd
     elif os.environ.get("ACC_LLAMA_TP") == "1":
         transformers.models.llama.modeling_llama.LlamaMLP = LlamaMLP
-    if os.getenv("XLA_USE_SPMD") == "1":
-        # use einsum in linear for SPMD TP/Ulysses.
-        transformers.models.llama.modeling_llama.LlamaAttention = LlamaAttention
-        transformers.models.llama.modeling_llama.LlamaDecoderLayer = LlamaDecoderLayer
+    # if os.getenv("XLA_USE_SPMD") == "1":
```
Comment (yitongh) [inline flashmodels/patch/patch.py]: do we need export XLA_USE_SPMD=1 when using spmd_fsdp?

Comment (baoleai) [reply flashmodels/patch/patch.py]: Yes, and here replace with `use_tp` args.

File: flashmodels/accelerators/acc_baichuan_accelerator.py (hunk 3 above)
Comment (baoleai) [reply flashmodels/accelerators/acc_baichuan_accelerator.py:19]: Tested and added support for all other models

File: examples/run.sh (hunk 2 above)
Comment (baoleai) [reply examples/run.sh]: good catch

File: examples/llama3_acc.sh (hunk 1 above)
Comment (baoleai) [reply examples/llama3_acc.sh]: deleted

**Review by yitongh — APPROVED** [review]

**Pull request merged**